File: Low Level Design (LLD)/Problems/02. ParkingLot/Java/service/ParkingLot.java
package service;

import model.*;
import strategy.ParkingFeeStrategy;

import java.time.LocalDateTime;
import java.util.*;

public class ParkingLot {
    private static ParkingLot instance;
    private final List<ParkingFloor> floors;
    private ParkingFeeStrategy feeStrategy;

    private ParkingLot() {
        floors = new ArrayList<>();
    }

    public static ParkingLot getInstance() {
        if (instance == null)
            instance = new ParkingLot();
        return instance;
    }

    public void setFeeStrategy(ParkingFeeStrategy strategy) {
        this.feeStrategy = strategy;
    }

    public void addFloor(ParkingFloor floor) {
        floors.add(floor);
    }

    public Ticket parkVehicle(Vehicle vehicle) {
        for (ParkingFloor floor : floors) {
            ParkingSlot slot = floor.getFreeSlot(vehicle);
            if (slot != null) {
                slot.assignVehicle(vehicle);
                return new Ticket(vehicle, slot);
            }
        }
        return null;
    }

    public double unparkVehicle(Ticket ticket) {
        LocalDateTime exitTime = LocalDateTime.now();
        double fee = feeStrategy.calculateFee(ticket.getVehicle(), ticket.getEntryTime(), exitTime);
        ticket.getSlot().removeVehicle(ticket.getVehicle());
        return fee;
    }
}

File: Low Level Design (LLD)/Problems/02. ParkingLot/Cpp/service/src/ParkingLot.cpp
#include "../include/ParkingLot.h"
#include <chrono>

ParkingLot* ParkingLot::instance = nullptr;

ParkingLot::ParkingLot() {}

ParkingLot* ParkingLot::getInstance() {
    if (instance == nullptr) {
        instance = new ParkingLot();
    }
    return instance;
}

void ParkingLot::setFeeStrategy(std::shared_ptr<ParkingFeeStrategy> strategy) {
    feeStrategy = strategy;
}

void ParkingLot::addFloor(std::shared_ptr<ParkingFloor> floor) {
    floors.push_back(floor);
}

std::shared_ptr<Ticket> ParkingLot::parkVehicle(Vehicle* vehicle) {
    for (auto& floor : floors) {
        ParkingSlot* slot = floor->getFreeSlot(vehicle);
        if (slot != nullptr) {
            slot->assignVehicle(vehicle);
            auto entryTime = std::chrono::system_clock::now();
            return std::make_shared<Ticket>(vehicle, slot, entryTime);
        }
    }
    return nullptr;
}

double ParkingLot::unparkVehicle(std::shared_ptr<Ticket> ticket) {
    auto exitTime = std::chrono::system_clock::now();
    double fee = feeStrategy->calculateFee(ticket->getVehicle(), ticket->getEntryTime(), exitTime);
    ticket->getSlot()->removeVehicle(ticket->getVehicle());
    return fee;
}
